_posts/2019-04-06-omakase-relay.markdown

```diff
@@ -0,0 +1,200 @@
+In REST clients, and GraphQL API clients like Apollo Client, you make a request and that request is passed through
+the React tree. E.g.
+
+{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}
+
+This means most components know more about the request than it probably needs, as it may be needed to pass on to
+the component's children. This can lead to data-duplication, or even worse, not knowing if you can delete or
+refactor a component.
+
+Data masking solves this by hiding data that the component didn't request. I've still yet to find the right visual
+abstraction, but I feel this just about pays for itself.
+
+{% include epic_img.html url="/images/omakase-relay/isolation.png" title="Relay isolation tree" style="width:100%;" %}
+
+You let Relay be responsible for consolidating all your fragments into a query, request it, and (mostly) passing it
+through your component hierarchy. This means Relay powered component can be safely changed and drastically reduces
+the chance for unintended consequences elsewhere.
+
+This isolation gives Artsy engineers the safety to work on projects with tens of contributors over long time
+periods without too much technical debt. The components we create are nearly all focused only on the data-driven
+aspects of rendering a GraphQL response into views.
+
+## Co-location
+
+Relay helped us move to one file representing everything a component needed. Effectively a single file now handles
+the styles, the actual view content hierarchy and the exact parts of the API it needs to render itself.
+
+<img src="/images/omakase-relay/co-location.png">
+
+In roughly that proportion too, though our most modern code uses the Artsy design system [Palette][palette] which
+drastically reduces the need for styling.
+
+## Community
+
+When we adopted Relay, there was no competition - we'd have just used the fetch API. Over time, Apollo came up and
+really put a considerable amount of effort into lowering the barriers to entry, and making it feasible to build
+complex apps easily.
+
+We did an audit last year of what it would take to re-create a lot of the infrastructure we use in Relay atop of
+the (much more popular) Apollo GraphQL eco-system and saw it was feasible but would require a considerable amount
+of work across many different plugins and tools. With Relay that's all packaged into one tool, works consistently
+and obviously doesn't have a scaling problem as Facebook have tens of thousands of Relay components.
+
+## Scale Safety
+
+Relay puts a lot of emphasis on ahead of time safety. The Relay compiler validates your queries against your
+GraphQL schema, we extended the compiler to create TypeScript types for the composed API fragments and there are
+strict naming systems enforced by the compiler. All of these guides engineers to build scalable codebases.
```

Comment:
I'm sure most people understand what "letting the compiler drive" means but maybe it would be cool to add a little screen cap showing how invalid GraphQL code fails at the compiler level and not at runtime? Apollo has similar validation features, but not without some setup.

Reply:
Done!